### PR TITLE
Allow caching of repodata from file:/// channels

### LINKED ISSUE #9661: Conda 4.7+ extremely slow "Collecting package metadata (repodata.json)" with local (file:///) channels
When installing or updating packages, conda gets stuck for an incredibly long time at the "Collecting package metadata (repodata.json)" step.

This appears to be a regression caused by conda 4.7 (and remains in 4.8) as this operation was quite fast in conda 4.6.

The slowdown only appears to occur when using (file:///) channels on a local or network filesystem. 

To test this I ran `python -m http.server 8000` in the channel parent directory and renamed channel URLs in `.condarc` from `file:///<DRIVE>:/<PATH>` to `http://localhost:8000/`. Commands then run in around the same time as they did in conda 4.6.

## Steps to Reproduce
`conda update conda` or any other conda update/install command

[conda-debug.txt](https://github.com/conda/conda/files/4156708/conda-debug.txt)
(note I exited (Ctrl+c) the command after a few minutes, normally it would run for an hour or more)
```
DEBUG conda.gateways.logging:set_verbosity(231): verbosity set to 2
DEBUG conda.core.solve:solve_final_state(223): solving prefix C:\Users\<USERNAME>\AppData\Local\Programs\Miniconda3
  specs_to_remove: frozenset()
  specs_to_add: frozenset({MatchSpec("conda")})
  prune: <auxlib._Null object at 0x0000024AA6792148>
DEBUG conda.core.package_cache_data:_check_writable(259): package cache directory 'C:\Users\<USERNAME>\AppData\Local\Programs\Miniconda3\pkgs' writable: True
DEBUG conda.core.subdir_data:_load(241): Local cache timed out for file:///X:/<LOCAL-MIRROR-PATH>/msys2/noarch/repodata.json at C:\Users\<USERNAME>\AppData\Local\Programs\Miniconda3\pkgs\cache\e8bdfbda.json
DEBUG conda.core.subdir_data:_load(241): Local cache timed out for file:///X:/<LOCAL-MIRROR-PATH>/msys2/win-64/repodata.json at C:\Users\<USERNAME>\AppData\Local\Programs\Miniconda3\pkgs\cache\270cd9d6.json
DEBUG conda.core.subdir_data:_load(241): Local cache timed out for file:///X:/<LOCAL-MIRROR-PATH>/main/win-64/repodata.json at C:\Users\<USERNAME>\AppData\Local\Programs\Miniconda3\pkgs\cache\a83aae50.json
DEBUG conda.core.subdir_data:_load(241): Local cache timed out for file:///X:/<LOCAL-MIRROR-PATH>/main/noarch/repodata.json at C:\Users\<USERNAME>\AppData\Local\Programs\Miniconda3\pkgs\cache\ce3ccb3a.json
INFO conda._vendor.auxlib.logz:stringify(144): request is 'None' for Response object with url file:///X:/<LOCAL-MIRROR-PATH>/msys2/noarch/repodata.json
DEBUG conda.core.subdir_data:fetch_repodata_remote_request(487):
<<FILE 200 None
< Content-Type: application/json
< Last-Modified: Tue, 13 Nov 2018 03:00:11 GMT
< Content-Length: 103
< Elapsed: 00:00.840189

{ 'info': {'subdir': 'noarch'},
  'packages': {},
  'removed': [],
  'repodata_version': 1}


DEBUG conda.core.subdir_data:_pickle_me(284): Saving pickled state for file:///X:/<LOCAL-MIRROR-PATH>/msys2/noarch/repodata.json at C:\Users\<USERNAME>\AppData\Local\Programs\Miniconda3\pkgs\cache\e8bdfbda.json
INFO conda._vendor.auxlib.logz:stringify(144): request is 'None' for Response object with url file:///X:/<LOCAL-MIRROR-PATH>/msys2/win-64/repodata.json
DEBUG conda.core.subdir_data:fetch_repodata_remote_request(487):
<<FILE 200 None
< Content-Type: application/json
< Last-Modified: Thu, 30 Jan 2020 00:18:12 GMT
< Content-Length: 227211
< Elapsed: 00:00.843118

{ 'info': {'subdir': 'win-64'},
  'packages': { 'm2-autoconf-2.69-3.tar.bz2': { 'build': '3',
                                                'build_number': 3,
                                                'depends': [ 'm2-bash',



INFO conda._vendor.auxlib.logz:stringify(144): request is 'None' for Response object with url file:///X:/<LOCAL-MIRROR-PATH>/main/noarch/repodata.json
DEBUG conda.core.subdir_data:_pickle_me(284): Saving pickled state for file:///X:/<LOCAL-MIRROR-PATH>/msys2/win-64/repodata.json at C:\Users\<USERNAME>\AppData\Local\Programs\Miniconda3\pkgs\cache\270cd9d6.json
DEBUG conda.core.subdir_data:fetch_repodata_remote_request(487):
<<FILE 200 None
< Content-Type: application/json
< Last-Modified: Thu, 30 Jan 2020 00:17:55 GMT
< Content-Length: 704400
< Elapsed: 00:00.841911

{ 'info': {'subdir': 'noarch'},
  'packages': { 'affine-2.1.0-pyh128a3a6_1.tar.bz2': { 'build': 'pyh128a3a6_1',
                                                       'build_number': 1,
                                                       'depends': ['py

```

```
conda.core.prefix_data.PrefixData.load,0.565156
conda.core.solve.Solver._collect_all_metadata,119.613876
conda.core.prefix_data.PrefixData.load,0.163001
conda.core.solve.Solver._collect_all_metadata,9.134097
conda.core.prefix_data.PrefixData.load,0.161001
conda.core.solve.Solver._collect_all_metadata,84.478895
```

## Environment Information
<details open><summary><code>`conda info`</code></summary><p>

```
     active environment : None
            shell level : 0
       user config file : C:\Users\<USERNAME>\.condarc
 populated config files : C:\Users\<USERNAME>\.condarc
          conda version : 4.8.1
    conda-build version : 3.18.11
         python version : 3.7.6.final.0
       virtual packages :
       base environment : C:\Users\<USERNAME>\AppData\Local\Programs\Miniconda3  (writable)
           channel URLs : file:///X:/<LOCAL-MIRROR-PATH>/main/win-64
                          file:///X:/<LOCAL-MIRROR-PATH>/main/noarch
                          file:///X:/<LOCAL-MIRROR-PATH>/msys2/win-64
                          file:///X:/<LOCAL-MIRROR-PATH>/msys2/noarch
          package cache : C:\Users\<USERNAME>\AppData\Local\Programs\Miniconda3\pkgs
                          C:\Users\<USERNAME>\.conda\pkgs
                          C:\Users\<USERNAME>\AppData\Local\conda\conda\pkgs
       envs directories : C:\Users\<USERNAME>\AppData\Local\Programs\Miniconda3\envs
                          C:\Users\<USERNAME>\.conda\envs
                          C:\Users\<USERNAME>\AppData\Local\conda\conda\envs
               platform : win-64
             user-agent : conda/4.8.1 requests/2.22.0 CPython/3.7.6 Windows/10 Windows/10.0.17763
          administrator : False
             netrc file : None
           offline mode : False
```
</p></details>

<details open><summary><code>`conda config --show-sources`</code></summary><p>

```
==> C:\Users\<USERNAME>\.condarc <==
auto_update_conda: False
auto_activate_base: False
channel_alias: file:///X:/<LOCAL-MIRROR-PATH>
channel_priority: disabled
channels:
  - defaults
default_channels:
  - main
  - msys2
show_channel_urls: True
use_only_tar_bz2: True
report_errors: False
```
</p></details>

<details><summary><code>`conda list --show-channel-urls`</code></summary><p>

```
# packages in environment at C:\Users\<USERNAME>\AppData\Local\Programs\Miniconda3:
#
# Name                    Version                   Build  Channel
asn1crypto                1.3.0                    py37_0    defaults
backports                 1.0                        py_2    defaults
backports.functools_lru_cache 1.6.1                      py_0    defaults
backports.tempfile        1.0                        py_1    defaults
backports.weakref         1.0.post1                  py_1    defaults
beautifulsoup4            4.8.2                    py37_0    defaults
bzip2                     1.0.8                he774522_0    defaults
ca-certificates           2019.11.27                    0    defaults
certifi                   2019.11.28               py37_0    defaults
cffi                      1.13.2           py37h7a1dbc1_0    defaults
chardet                   3.0.4                 py37_1003    defaults
click                     7.0                        py_0    defaults
conda                     4.8.1                    py37_0    defaults
conda-build               3.18.11                  py37_0    defaults
conda-package-handling    1.6.0            py37h62dcd97_0    defaults
conda-verify              3.4.2                      py_1    defaults
console_shortcut          0.1.1                         3    https://repo.anaconda.com/pkgs/main
constructor               3.0.0                    py37_0    defaults
cryptography              2.8              py37h7a1dbc1_0    defaults
filelock                  3.0.12                     py_0    defaults
freetype                  2.9.1                ha9979f8_1    defaults
future                    0.18.2                   py37_0    defaults
glob2                     0.7                        py_0    defaults
idna                      2.8                      py37_0    https://repo.anaconda.com/pkgs/main
jinja2                    2.10.3                     py_0    defaults
jpeg                      9b                   hb83a4c4_2    defaults
libarchive                3.3.3                h0643e63_5    defaults
libiconv                  1.15                 h1df5818_7    defaults
liblief                   0.9.0                ha925a31_2    defaults
libpng                    1.6.37               h2a8f88b_0    defaults
libtiff                   4.1.0                h56a325e_0    defaults
libxml2                   2.9.9                h464c3ec_0    defaults
lz4-c                     1.8.1.2              h2fa13f4_0    defaults
lzo                       2.10                 h6df0209_2    defaults
m2-msys2-runtime          2.5.0.17080.65c939c               3    defaults
m2-patch                  2.7.5                         2    defaults
markupsafe                1.1.1            py37he774522_0    defaults
menuinst                  1.4.16           py37he774522_0    https://repo.anaconda.com/pkgs/main
msys2-conda-epoch         20160418                      1    defaults
nsis                      3.01                          8    defaults
olefile                   0.46                     py37_0    defaults
openssl                   1.1.1d               he774522_3    defaults
pillow                    7.0.0            py37hcc1f983_0    defaults
pip                       20.0.2                   py37_0    defaults
pkginfo                   1.5.0.1                  py37_0    defaults
powershell_shortcut       0.0.1                         2    https://repo.anaconda.com/pkgs/main
psutil                    5.6.7            py37he774522_0    defaults
py-lief                   0.9.0            py37ha925a31_2    defaults
pycosat                   0.6.3            py37he774522_0    defaults
pycparser                 2.19                       py_0    defaults
pyopenssl                 19.1.0                   py37_0    defaults
pysocks                   1.7.1                    py37_0    defaults
python                    3.7.6                h60c2a47_2    defaults
python-dateutil           2.8.1                      py_0    defaults
python-libarchive-c       2.8                     py37_13    defaults
pytz                      2019.3                     py_0    defaults
pywin32                   227              py37he774522_1    defaults
pyyaml                    5.2              py37he774522_0    defaults
requests                  2.22.0                   py37_1    defaults
ruamel_yaml               0.15.87          py37he774522_0    defaults
setuptools                45.1.0                   py37_0    defaults
six                       1.14.0                   py37_0    defaults
soupsieve                 1.9.5                    py37_0    defaults
sqlite                    3.30.1               he774522_0    defaults
tk                        8.6.8                hfa6e2cd_0    defaults
tqdm                      4.42.0                     py_0    defaults
urllib3                   1.25.8                   py37_0    defaults
vc                        14.1                 h0510ff6_4    https://repo.anaconda.com/pkgs/main
vs2015_runtime            14.16.27012          hf0eaf9b_1    defaults
wheel                     0.33.6                   py37_0    defaults
win_inet_pton             1.1.0                    py37_0    https://repo.anaconda.com/pkgs/main
wincertstore              0.2                      py37_0    https://repo.anaconda.com/pkgs/main
xz                        5.2.4                h2fa13f4_4    defaults
yaml                      0.1.7                hc54c509_2    https://repo.anaconda.com/pkgs/main
zlib                      1.2.11               h62dcd97_3    defaults
zstd                      1.3.7                h508b16e_0    defaults
```
</p></details>

### PR DESCRIPTION
**Pull request by LorcanHamill**
(This is the second attempt at this, sorry for any inconvenience.)

As discussed in issue #9661 the performance of "file:///" channels is very poor in some circumstances with the latest conda versions. This seems to be because the solver requests repodata about directories more often than it used to. There is a caching mechanism for repodata but it is explicitly disabled for local channels; this PR removes that check so that the caching mechanism works for file:/// channels in the same way as HTTP channels.

That broke some unit tests, because of the way that an environment variable can be set to force conda to use only .tar.bz2 files, not .conda files. The tests try both settings, which changes the repodata that is fetched and should therefore be cached. The tests fail because the wrong version of the data appears in the cache. This could never happen in production as the cache is in memory. So it starts out empty when the conda command is invoked, and the command will only see one setting of the environment variable, so the cache will only ever have the appropriate data. To fix the tests, I added a method to clear the cache and invoked it before each test.

This change makes a huge difference to my company's use of conda. We use large conda channels on relatively slow NFS volumes, and without this change the creation of a new environment takes minutes when it previously took a few seconds.

Fixes #9661

### CODE REVIEW

**Comment by LorcanHamill** [comment]
Okay, this is a new PR that just adds a class method that is not (yet) called from anywhere.  And yet lots of tests are broken.   I would be very grateful if someone could explain what I've done wrong..?

**Comment by LorcanHamill** [comment]
To be clear, this draft PR is just meant to be a starting point; I suspected that the tests might fail, and they did, but I can't see how my code change could have caused that.   When the tests are passing I'll update the PR to have a full implementation of the new caching behaviour.

**Comment by LorcanHamill** [comment]
I've been away from this for some time but hope to start work again on it shortly.  I've just signed the CLA again; I think it may not have worked previously because my GitHub profile somehow didn't include my email address.   Hopefully the CLA process will work this time...

**Comment by LorcanHamill** [comment]
Could someone please check why the "CLA-Signed" check is failing on this PR?  I've signed the CLA at least three times, and the bot eventually marked a previous PR (#9677) with the "cla-signed" label on March 3. Thanks!

**Comment by jjhelmus** [comment]
@cla-bot check

**Comment by jjhelmus** [comment]
> Could someone please check why the "CLA-Signed" check is failing on this PR? I've signed the CLA at least three times, and the bot eventually marked a previous PR (#9677) with the "cla-signed" label on March 3. Thanks!

This process is manually.  I've added you to the signed list and summoned the bot.

**Comment by LorcanHamill** [comment]
> This process is manually. I've added you to the signed list and summoned the bot.

Ah, I see - I had assumed it was an automated process.  Thanks, @jjhelmus !

**Comment by LorcanHamill** [comment]
As far as I can tell, this PR is ready. The AppVeyor build didn't fail - it was killed for exceeding the 1 hour time limit.

The issue that this fixes - #9661 - is preventing our company from upgrading to conda 4.8, so I'd really like to see this merged soon.  Please let me know if there's anything I can do to help that happen.  (This is my first conda PR, so I'm not sure of the procedure.)   Thanks!

**Comment by LorcanHamill** [comment]
I don't understand what's happened.   All checks were passing on this PR, except that the AppVeyor build was timing out.   I pushed a couple of empty commits to trigger re-runs of the checks, and now some tests are failing.   Guess I may have rebased on top of other changes that broke those tests?  Or is something else going on?   If anyone could enlighten me, that would be great!

**Comment by aldanor** [comment]
@msarahan Could you please take a look at this?

**Comment by LorcanHamill** [comment]
Ok, this PR is passing all checks again, apart from AppVeyor which appears to be broken?

Is there anything else that I need to do, to get this processed?   If not, can it **_please_** be reviewed and merged?

I appreciate that everyone is busy, but this is essentially a one-line code change, with the additional tweaks that are required to keep the tests passing.   It is very important to our company; we can't possibly upgrade to conda 4.8 without this change or something similar; the performance in our environment is so bad as to be unusable.   We're not the only ones affected, either  - see #9661 and comments on #9677 from @carloswanguw

It's almost three months since I submitted the first version of this change, and I'd really like to see it accepted soon.

**Comment by msarahan** [comment]
I am no longer employed by Anaconda and can't merge anything. I'm not sure about this pr. How does the cache for local repodata ever get invalidated? From what I can tell, it doesn't. That will cause immense problems to anyone who builds packages.

**Comment by LorcanHamill** [comment]
> I am no longer employed by Anaconda and can't merge anything. I'm not sure about this pr. How does the cache for local repodata ever get invalidated? From what I can tell, it doesn't. That will cause immense problems to anyone who builds packages.

Thanks for the response.  I was unaware of the change in your situation, I hope that's a positive development for you.

The cache is an in-memory one that only lives for the duration of the process and is now handled in the same way for file:// and https:// URLs.    Does the metadata for file URLs get used during package building?  And does it _change_ during package building, after it has been read, and during the lifetime of a single conda process?   (To be clear, those are not rhetorical questions, I don't know the answers.)  In any case, I would have assumed that issues like that would be caught by one of the many unit and integration tests?

**Comment by msarahan** [comment]
> Does the metadata for file URLs get used during package building? And does it change during package building, after it has been read, and during the lifetime of a single conda process?

Yes, unfortunately. Conda-build loads this the first time it creates the build or host environments, and then it needs to reload it once the new package has been built so that the test environment creation can use the new package. Conda-build could add a call to your new function here that clears things, but this change would break all past versions of conda-build.

**Comment by msarahan** [comment]
Perhaps a cache invalidation scheme based on checking mtime as part of the cache key might work.

**Comment by LorcanHamill** [comment]
> Perhaps a cache invalidation scheme based on checking mtime as part of the cache key might work.

Thanks, I'll experiment with that.  I'm concerned that checking mtime will be almost as slow as reading the data on these network shares, but perhaps not; it's worth trying, at least.

I confess that I know nothing about how conda-build interacts with this code in conda itself.   Is there perhaps some other interaction between the two of them during package building that could be used to invalidate the cache?

I'll try the mtime thing anyway, and hopefully push an update soon.

**Comment by LorcanHamill** [comment]
Just FYI, I'm returning to this after a long break.  I've converted it back into a draft while I investigate the issue related to conda-build that was identified by @msarahan

**Comment by LorcanHamill** [comment]
I believe this PR is now ready for review.

The problem identified by @msarahan was real.   When I tried to run the unit tests in conda-build against this modified version of conda, many of the tests failed.   (I killed it after about 10 or 20 tests had failed, so I don't know how many were affected.)

So I've implemented the suggested solution, comparing the modification time of the JSON file with the timestamp on the cache entry, and discarding the cache entry if it's older.   With this version, the conda-build tests all pass (except one, which also fails when built against the master branch of conda, so that's something unrelated).

One puzzling issue remains.  One of my new unit tests - `tests/core/test_subdir_data.py::test_metadata_cache_works` - is failing in AppVeyor.   It passes consistently in circleci for both py27 and py37, and also when I run it locally.   But it always fails in AppVeyor, for both py27 and py36.   This test just relies on files having a valid modification time and the "time.time()" method returning the current time.   Is it possible that in the AppVeyor environment, system clocks are not synchronized properly between hosts and NFS servers?    That could account for the test failing.

In any case, since the same test passes reliably in circleci, I hope the appveyor failure will not block review of this PR.

As before, I'd be very grateful if this PR could be reviewed (and hopefully approved) as soon as possible.   We have many file-based conda channels in our environment, and this makes a huge difference in the usablility of conda 4.8.      We have one package with many dependencies - let's call it `foo`.  I measured the time it takes to run this command:

`conda create -d -n foo foo`

(So it solves the dependencies to create an environment, but doesn't actually create it.)

With the standard conda 4.8.3, that took almost eight minutes (7:49.06).  With this caching mechanism in place, it took less than 15 seconds (0.14.49).   I think those figures show that this change is very worthwhile.

**Comment by LorcanHamill** [comment]
This now passes all tests.  For some reason AppVeyor is marking the py27 run of the integration tests as a failure,  None of the tests actually failed; there was one XFAIL, but that hardly causes the job to fail?

Anyway, this is once again ready for review.

**Comment by aldanor** [comment]
@mcg1969 @chenghlee @jjhelmus Could this be re-reviewed please? This is a major regression so would be nice to have it finally patched.

**Comment by LorcanHamill** [comment]
This PR has been awaiting review for over a week now.  Is there anything else I can do to try to progress this?

I'm not sure how to request review other than by tagging people in a comment here, so I'll mention everyone who recently approved or merged PRs:

@chenghlee @mcg1969 @angloyna @jjhelmus 

Apologies if that looks like I'm spamming people, but I don't know what else to do.  There doesn't seem to be any other forum to discuss this sort of thing?  (The "conda" mailing list, mentioned in some docs that I've seen, appears to be dead?)

**Comment by mcg1969** [comment]
@LorcanHamill as you can see we have a fair amount of backlog here. I assure you we're going to get to it, and in fact we are in the process of making some internal adjustments so we can devote more resources to clearing this backlog. That's all I can assure you of now, except to assure you I'm really not trying to brush you or this useful PR off. (And to be honest, even if it were merged, you'd then be waiting for a release to be tagged, too.)

**Comment by LorcanHamill** [comment]
Thanks for the response, @mcg1969 .  It's the not knowing that is most frustrating, as I'm sure you understand.   I appreciate that it's not always possible to keep up with incoming PRs.   As a matter of interest, is there a schedule somewhere for releases, with cut-off dates for merges and so on?   It would help with our planning if we have some idea of when this would eventually make it into a release, if all goes well.

**Comment by mcg1969** [comment]
I think that's a good suggestion @LorcanHamill ... as it turns out there are active discussions in the conda community around the larger questions of governance for this project. I must admit I'm not involved in those discussions so I can't tell you when or what but I'm genuinely positive changes are coming.

**Comment by LorcanHamill** [comment]
I gather this PR has missed the cut for 4.8.4?   Is there something I need to do now, to keep this PR alive?

**Review by chenghlee — APPROVED** [review]

**Comment by chenghlee** [comment]
Closing & reopening this PR to force new CI checks to run; once those pass, I'll merge this PR.

**Comment by LorcanHamill** [comment]
Great to see this merged - thanks, @chenghlee !

**Comment by haywardlam** [comment]
My company is using the latest conda 4.9.2 and experiencing the same problem with network drive based channels. Any idea what version of conda will have this fix?

@LorcanHamill  Really appreciate your work on this.

**Comment by LorcanHamill** [comment]
This was included in conda 4.9.0, so it should be in the version you're
using.

https://github.com/conda/conda/blob/master/CHANGELOG.md#490-2020-10-19

You're welcome :)

**Comment by haywardlam** [comment]
@LorcanHamill I see it in the changelog 4.9.0. However, I am using 4.9.2 (latest Anaconda install) with our file:// network drive and conda is still extremely slow. Is there any configuration in .condarc required to use this caching feature?

**Comment by LorcanHamill** [comment]
No, there is no special configuration required, it should just happen.
Maybe your cache folder itself is on an NFS drive?

**Pull request merged**
